File: src/main/java/aporia/su/modules/impl/combat/NoFriendDamage.java
package aporia.su.modules.impl.combat;

import lombok.AccessLevel;
import lombok.experimental.FieldDefaults;
import aporia.su.util.events.api.EventHandler;
import aporia.su.util.events.impl.entity.InteractEntityEvent;
import aporia.su.modules.module.ModuleStructure;
import aporia.su.modules.module.category.ModuleCategory;
import aporia.su.util.user.repository.friend.FriendUtils;

@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class NoFriendDamage extends ModuleStructure {

    public NoFriendDamage() {
        super("NoFriendDamage", "No Friend Damage", ModuleCategory.COMBAT);
    }

    @EventHandler
    public void onAttack(InteractEntityEvent e) {
        e.setCancelled(FriendUtils.isFriend(e.getEntity()));
    }
}

File: src/so/aporia/module/impl/combat/NoFriendDamage.kt
package so.aporia.module.impl.combat

import net.minecraft.client.Minecraft
import net.minecraft.network.protocol.game.ServerboundInteractPacket
import net.minecraft.world.entity.player.Player
import so.aporia.module.Category
import so.aporia.module.Module
import so.aporia.utils.events.EventBus
import so.aporia.utils.events.EventHandler
import so.aporia.utils.events.impl.PacketEvent
import so.aporia.utils.user.friend.FriendManager

class NoFriendDamage : Module("NoFriendDamage", Category.COMBAT) {
    companion object {
        @JvmField val mc = Minecraft.getInstance()
        private val entityIdField by lazy {
            val f = ServerboundInteractPacket::class.java.getDeclaredField("entityId")
            f.isAccessible = true
            f
        }
    }

    override fun onEnable() {
        EventBus.register(this)
    }

    override fun onDisable() {
        EventBus.unregister(this)
    }

    @EventHandler
    fun onPacket(event: PacketEvent) {
        if (event.direction() != PacketEvent.Direction.OUTBOUND) return
        if (mc.player == null || mc.level == null) return

        val packet = event.packet()
        if (packet is ServerboundInteractPacket) {
            val entityId = entityIdField.getInt(packet)
            val entity = mc.level!!.getEntity(entityId)
            if (entity is Player && FriendManager.isFriend(entity.name.string)) {
                event.cancel()
            }
        }
    }
}
